Subject: tst-timer4 fails
From: Olaf Hering <olaf@suse.de>
References: 43799 - LTC10642
Patch-mainline: Not yet

tst-timer4 from 32bit glibc on PPC64 fails with:

clock_gettime returned timespec = { 1092395908, 871187000 }
clock_getres returned timespec = { 0, 999848 }
+*** sig1_sigval.sival_ptr (nil) != 0xffffd1c0
+*** sig2_sigval.sival_ptr 0 != 163
!!! second timer_gettime timer_none returned it_interval 0.100984648

This is due to a incomplete structure in compat_linux for SI_TIMER.

Signed-off-by: Hannes Reinecke <hare@suse.de>

--- linux-2.6.5/arch/s390/kernel/compat_linux.h.orig	2004-11-03 15:22:32.304406099 +0100
+++ linux-2.6.5/arch/s390/kernel/compat_linux.h	2004-11-03 15:58:57.714406099 +0100
@@ -50,9 +50,10 @@
 
 		/* POSIX.1b timers */
 		struct {
-			unsigned int	_timer1;
-			unsigned int	_timer2;
-                
+			timer_t _tid;			/* timer id */
+			int _overrun;			/* overrun count */
+			compat_sigval_t _sigval;	/* same as below */
+			int _sys_private;		/* not to be passed to user */
 		} _timer;
 
 		/* POSIX.1b signals */
@@ -98,6 +99,8 @@
 #define si_addr		_sifields._sigfault._addr
 #define si_band		_sifields._sigpoll._band
 #define si_fd		_sifields._sigpoll._fd    
+#define si_tid		_sifields._timer._tid
+#define si_overrun	_sifields._timer._overrun
 
 /* asm/sigcontext.h */
 typedef union
--- linux-2.6.5/arch/s390/kernel/compat_signal.c.orig	2004-11-03 15:23:10.524406099 +0100
+++ linux-2.6.5/arch/s390/kernel/compat_signal.c	2004-11-03 15:58:20.794406099 +0100
@@ -94,10 +94,14 @@
 					  &to->si_addr);
 			break;
 		case __SI_POLL >> 16:
-		case __SI_TIMER >> 16:
 			err |= __put_user(from->si_band, &to->si_band);
 			err |= __put_user(from->si_fd, &to->si_fd);
 			break;
+		case __SI_TIMER >> 16:
+			err |= __put_user(from->si_tid, &to->si_tid);
+			err |= __put_user(from->si_overrun, &to->si_overrun);
+			err |= __put_user(from->si_int, &to->si_int);
+			break;
 		default:
 			break;
 		}
@@ -139,10 +143,14 @@
 			to->si_addr = (void *)(u64) (tmp & PSW32_ADDR_INSN);
 			break;
 		case __SI_POLL >> 16:
-		case __SI_TIMER >> 16:
 			err |= __get_user(to->si_band, &from->si_band);
 			err |= __get_user(to->si_fd, &from->si_fd);
 			break;
+		case __SI_TIMER >> 16:
+			err |= __get_user(to->si_tid, &from->si_tid);
+			err |= __get_user(to->si_overrun, &from->si_overrun);
+			err |= __get_user(to->si_int, &from->si_int);
+			break;
 		default:
 			break;
 		}
